src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -277,21 +287,355 @@ openLevels reg hfs hbio TableConfig{..} sessionRoot levels =
+            printf "Incompatible version found. Version %s is not backwards \
+                   \compatible with version %s : %s"
+                   (prettySnapshotVersion currentSnapshotVersion)
+                   (prettySnapshotVersion version)
+                   errMsg
+      Versioned <$> decode' version
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: Versioning
+-------------------------------------------------------------------------------}
+
+instance Encode SnapshotVersion where
+  encode ver =
+         encodeListLen 2
+      <> encodeWord (major ver)
+      <> encodeWord (minor ver)
+
+instance Decode SnapshotVersion where
+  decode = do
+      _ <- decodeListLenOf 2
+      majo <- decodeWord
+      mino <- decodeWord
+      case fromMajorMinor majo mino of
+        Nothing -> fail ("Unknown version number: " <> prettyVersion majo mino)
+        Just version -> pure version
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: SnapshotMetaData
+-------------------------------------------------------------------------------}
+
+-- SnapshotMetaData
+
+instance Encode SnapshotMetaData where
+  encode (SnapshotMetaData label config levels) =
+         encodeListLen 3
+      <> encode label
+      <> encode config
+      <> encode levels
+
+instance DecodeVersioned SnapshotMetaData where
+  decode' ver@V0_0 = do
+      _ <- decodeListLenOf 3
+      SnapshotMetaData <$> decode' ver <*> decode' ver <*> decode' ver
+
+-- SnapshotLabel
+
+instance Encode SnapshotLabel where
+  encode (SnapshotLabel s) = encodeString s
+
+instance DecodeVersioned SnapshotLabel where
+  decode' V0_0 = SnapshotLabel <$> decodeString
+
+{-------------------------------------------------------------------------------
+  Encoding and decoding: TableConfig
 -------------------------------------------------------------------------------}
 
-deriving stock instance Read MergePolicyForLevel
-deriving newtype instance Read RunNumber
-deriving stock instance Read Merge.Level
+-- TableConfig
+
+instance Encode TableConfig where
+  encode config =
+         encodeListLen 7
+      <> encode mergePolicy
+      <> encode sizeRatio
+      <> encode writeBufferAlloc
+      <> encode bloomFilterAlloc
+      <> encode fencePointerIndex
+      <> encode diskCachePolicy
+      <> encode mergeSchedule
+    where
+      TableConfig
+        mergePolicy
+        sizeRatio
+        writeBufferAlloc
+        bloomFilterAlloc
+        fencePointerIndex
+        diskCachePolicy
+        mergeSchedule
+        = config
+
+instance DecodeVersioned TableConfig where
+  decode' v@V0_0 = do
+      _ <- decodeListLenOf 7
+      TableConfig
+        <$> decode' v <*> decode' v <*> decode' v
+        <*> decode' v <*> decode' v <*> decode' v
+        <*> decode' v
+
+-- MergePolicy
+
+instance Encode MergePolicy where
+  encode MergePolicyLazyLevelling = encodeWord 0
```

Comment:
This can't be extended to variants with extra fields (like any of the parametrised merge policies in the papers). It's unlikely we'll need it, but on the other hand it's easy to make this extensible now and it seems like we allow future extensions in some other places already. So use your own judgement.

Reply:
Good point. It should at least have a list length field for the reason you mentioned

Reply:
Add list lengths in a few more places